models/QBMigrationManager.cfc

```diff
@@ -0,0 +1,158 @@
+	property name="schema";
+	property name="useTransactions" default="true";
+
+	boolean function isReady() {
+		return isMigrationTableInstalled();
+	}
+
+	function install( runAll = false ) {
+		if ( isMigrationTableInstalled() ) {
+			return;
+		}
+
+		var schema = wirebox.getInstance( "SchemaBuilder@qb" ).setGrammar( wirebox.getInstance( defaultGrammar ) );
+
+		schema.create(
+			getMigrationsTable(),
+			function( table ) {
+				table.string( "name", 190 ).primaryKey();
+				table.datetime( "migration_ran" );
+			},
+			{ datasource : getDatasource() }
+		);
+	}
+
+	public void function uninstall() {
+		if ( !isMigrationTableInstalled() ) {
+			return;
+		}
+
+		queryExecute(
+			"DROP TABLE #getMigrationsTable()#",
+			{},
+			{ datasource : getDatasource() }
+		);
+	}
+
+	public void function reset() {
+		var schema = wirebox.getInstance( "SchemaBuilder@qb" ).setGrammar( wirebox.getInstance( defaultGrammar ) );
+		schema.dropAllObjects( options = { datasource : getDatasource() }, schema = getSchema() );
+	}
+
+	array function findProcessed() {
+		return queryExecute(
+			"
+                SELECT name
+                FROM #getMigrationsTable()#
+            ",
+			{},
+			{ datasource : getDatasource() }
+		).reduce( function( result, row ) {
+				result.append( row );
+				return result;
+			}, [] )
+			.map( function( row ) {
+				return row.name;
+			} );
+	}
+
+
+	boolean function isMigrationRan( componentName ) {
+		var processed = findProcessed();
+		return processed.contains( componentName );
+	}
+
+
+	private void function logMigration( direction, componentName ) {
+		if ( direction == "up" ) {
+			queryExecute(
+				"INSERT INTO #getMigrationsTable()# VALUES ( :name, :time )",
+				{
+					name : componentName,
+					time : {
+						value     : now(),
+						cfsqltype : "CF_SQL_TIMESTAMP"
+					}
+				},
+				{ datasource : getDatasource() }
+			);
+		} else {
+			queryExecute(
+				"DELETE FROM #getMigrationsTable()# WHERE name = :name",
+				{ name : componentName },
+				{ datasource : getDatasource() }
+			);
+		}
+	}
+
+
+	public void function runMigration(
+		direction,
+		migrationStruct,
+		postProcessHook,
+		preProcessHook
+	) {
+		install();
+
+		var migrationRan = isMigrationRan( migrationStruct.componentName );
+
+		if ( migrationRan && direction == "up" ) {
+			throw( "Cannot run a migration that has already been ran." );
+		}
+
+		if ( !migrationRan && direction == "down" ) {
+			throw( "Cannot rollback a migration if it hasn't been ran yet." );
+		}
+
+		var migration = wirebox.getInstance( migrationStruct.componentPath );
+
+		var schema = wirebox.getInstance( "SchemaBuilder@qb" ).setGrammar( wirebox.getInstance( defaultGrammar ) );
+
+		var query = wirebox.getInstance( "QueryBuilder@qb" ).setGrammar( wirebox.getInstance( defaultGrammar ) );
+
+		preProcessHook( migrationStruct );
+
+		$transactioned( function() {
+			invoke(
+				migration,
+				direction,
+				[ schema, query ]
```

Comment:
I like this because each transaction should be able to provide it's own arguments to their migration files.